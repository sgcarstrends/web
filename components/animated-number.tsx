"use client";

import { useEffect } from "react";
import { motion, useSpring, useTransform } from "framer-motion";

interface Props {
  value: number;
  from?: number;
}

export const AnimatedNumber = ({ value, from = 0 }: Props) => {
  const spring = useSpring(from, { mass: 0.8, stiffness: 75, damping: 15 });
  const display = useTransform(spring, (current) =>
    Math.round(current).toLocaleString(),
  );

  useEffect(() => {
    spring.set(value);
  }, [spring, value]);

  return <motion.span>{display}</motion.span>;
};
