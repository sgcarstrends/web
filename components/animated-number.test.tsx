import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { AnimatedNumber } from "./animated-number";

describe("AnimatedNumber", () => {
  it("should render without crashing", () => {
    const { container } = render(<AnimatedNumber value={100} />);
    expect(container).toMatchSnapshot();
  });

  it("should display the initial value when from is specified", () => {
    render(<AnimatedNumber value={100} from={50} />);
    expect(screen.getByText("50")).toBeVisible();
  });

  it("should animate to the final value", async () => {
    render(<AnimatedNumber value={100} from={0} />);

    await new Promise((resolve) => setTimeout(resolve, 500)); // Adjust timeout as needed

    await waitFor(() => {
      expect(screen.getByText("100")).toBeVisible();
    });
  });

  it("should format the number with toLocaleString", async () => {
    render(<AnimatedNumber value={1000} from={0} />);

    await waitFor(() => {
      expect(screen.getByText("1,000")).toBeVisible();
    });
  });
});
