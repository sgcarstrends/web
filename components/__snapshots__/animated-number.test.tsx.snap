// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AnimatedNumber > should render without crashing 1`] = `
<div>
  <span>
    0
  </span>
</div>
`;
